# Answering Business Questions Using SQL

The first thing I'll do is look through the database to see what tables are in it. From there, I'll inspect some of the tables to see how I can use them.

In [1]:
%%capture
%load_ext sql
%sql sqlite:///chinook.db

In [2]:
%%sql
SELECT
    name,
    type
FROM sqlite_master
WHERE type IN ("table", "view");

 * sqlite:///chinook.db
Done.


[('album', 'table'),
 ('artist', 'table'),
 ('customer', 'table'),
 ('employee', 'table'),
 ('genre', 'table'),
 ('invoice', 'table'),
 ('invoice_line', 'table'),
 ('media_type', 'table'),
 ('playlist', 'table'),
 ('playlist_track', 'table'),
 ('track', 'table'),
 ('invoice_tracks', 'view'),
 ('album_tracks', 'view')]

In [3]:
%%sql
SELECT *
FROM album
LIMIT 5;

 * sqlite:///chinook.db
Done.


[(1, 'For Those About To Rock We Salute You', 1),
 (2, 'Balls to the Wall', 2),
 (3, 'Restless and Wild', 2),
 (4, 'Let There Be Rock', 1),
 (5, 'Big Ones', 3)]

In [4]:
%%sql
SELECT *
FROM artist
LIMIT 5;

 * sqlite:///chinook.db
Done.


[(1, 'AC/DC'),
 (2, 'Accept'),
 (3, 'Aerosmith'),
 (4, 'Alanis Morissette'),
 (5, 'Alice In Chains')]

In [5]:
%%sql
SELECT *
FROM customer
LIMIT 5;

 * sqlite:///chinook.db
Done.


[(1, 'Luís', 'Gonçalves', 'Embraer - Empresa Brasileira de Aeronáutica S.A.', 'Av. Brigadeiro Faria Lima, 2170', 'São José dos Campos', 'SP', 'Brazil', '12227-000', '+55 (12) 3923-5555', '+55 (12) 3923-5566', 'luisg@embraer.com.br', 3),
 (2, 'Leonie', 'Köhler', None, 'Theodor-Heuss-Straße 34', 'Stuttgart', None, 'Germany', '70174', '+49 0711 2842222', None, 'leonekohler@surfeu.de', 5),
 (3, 'François', 'Tremblay', None, '1498 rue Bélanger', 'Montréal', 'QC', 'Canada', 'H2G 1A7', '+1 (514) 721-4711', None, 'ftremblay@gmail.com', 3),
 (4, 'Bjørn', 'Hansen', None, 'Ullevålsveien 14', 'Oslo', None, 'Norway', '0171', '+47 22 44 22 22', None, 'bjorn.hansen@yahoo.no', 4),
 (5, 'František', 'Wichterlová', 'JetBrains s.r.o.', 'Klanova 9/506', 'Prague', None, 'Czech Republic', '14700', '+420 2 4172 5555', '+420 2 4172 5555', 'frantisekw@jetbrains.com', 4)]

In [6]:
%%sql
SELECT *
FROM invoice
LIMIT 5;

 * sqlite:///chinook.db
Done.


[(1, 18, '2017-01-03 00:00:00', '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', 15.84),
 (2, 30, '2017-01-03 00:00:00', '230 Elgin Street', 'Ottawa', 'ON', 'Canada', 'K2P 1L7', 9.9),
 (3, 40, '2017-01-05 00:00:00', '8, Rue Hanovre', 'Paris', 'None', 'France', '75002', 1.98),
 (4, 18, '2017-01-06 00:00:00', '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', 7.92),
 (5, 27, '2017-01-07 00:00:00', '1033 N Park Ave', 'Tucson', 'AZ', 'USA', '85719', 16.83)]

# Finding Most Popular Genres in USA

The first thing we'll need to do is find out what the most popular genres are by both total track count and by percentage. From there, we'll work backward to only include tracks from customers from the US. In order to find the percentage, we'll need to know the total amount of tracks as well.

In [7]:
%%sql
SELECT COUNT(track_id)
FROM track;

 * sqlite:///chinook.db
Done.


[(3503,)]

There are 3503 tracks in this table.

In [8]:
%%sql
SELECT
    g.name genre,
    COUNT(t.track_id) track_count,
    COUNT(t.track_id)/3503. track_pct
FROM genre g
INNER JOIN track t ON t.genre_id = g.genre_id
GROUP BY 1

 * sqlite:///chinook.db
Done.


[('Alternative', 40, 0.011418783899514702),
 ('Alternative & Punk', 332, 0.09477590636597202),
 ('Blues', 81, 0.02312303739651727),
 ('Bossa Nova', 15, 0.004282043962318013),
 ('Classical', 74, 0.0211247502141022),
 ('Comedy', 17, 0.004852983157293748),
 ('Drama', 64, 0.018270054239223524),
 ('Easy Listening', 24, 0.006851270339708821),
 ('Electronica/Dance', 30, 0.008564087924636025),
 ('Heavy Metal', 28, 0.00799314872966029),
 ('Hip Hop/Rap', 35, 0.009991435912075365),
 ('Jazz', 130, 0.03711104767342278),
 ('Latin', 579, 0.1652868969454753),
 ('Metal', 374, 0.10676562946046246),
 ('Opera', 1, 0.00028546959748786756),
 ('Pop', 48, 0.013702540679417643),
 ('R&B/Soul', 61, 0.01741364544675992),
 ('Reggae', 58, 0.016557236654296318),
 ('Rock', 1297, 0.3702540679417642),
 ('Rock And Roll', 12, 0.0034256351698544107),
 ('Sci Fi & Fantasy', 26, 0.007422209534684556),
 ('Science Fiction', 13, 0.003711104767342278),
 ('Soundtrack', 43, 0.012275192691978304),
 ('TV Shows', 93, 0.02654867256637168),
 ('World', 28, 0.00799314872966029)]

Okay, now we've got the exact chart we need, but this is for customers all over the world. Now we need to find a way to only count US customers.

In [9]:
%%sql
SELECT *
FROM customer
WHERE country = "USA"
LIMIT 10;

 * sqlite:///chinook.db
Done.


[(16, 'Frank', 'Harris', 'Google Inc.', '1600 Amphitheatre Parkway', 'Mountain View', 'CA', 'USA', '94043-1351', '+1 (650) 253-0000', '+1 (650) 253-0000', 'fharris@google.com', 4),
 (17, 'Jack', 'Smith', 'Microsoft Corporation', '1 Microsoft Way', 'Redmond', 'WA', 'USA', '98052-8300', '+1 (425) 882-8080', '+1 (425) 882-8081', 'jacksmith@microsoft.com', 5),
 (18, 'Michelle', 'Brooks', None, '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', '+1 (212) 221-3546', '+1 (212) 221-4679', 'michelleb@aol.com', 3),
 (19, 'Tim', 'Goyer', 'Apple Inc.', '1 Infinite Loop', 'Cupertino', 'CA', 'USA', '95014', '+1 (408) 996-1010', '+1 (408) 996-1011', 'tgoyer@apple.com', 3),
 (20, 'Dan', 'Miller', None, '541 Del Medio Avenue', 'Mountain View', 'CA', 'USA', '94040-111', '+1 (650) 644-3358', None, 'dmiller@comcast.com', 4),
 (21, 'Kathy', 'Chase', None, '801 W 4th Street', 'Reno', 'NV', 'USA', '89503', '+1 (775) 223-7665', None, 'kachase@hotmail.com', 5),
 (22, 'Heather', 'Leacock', None, '120 S Orange Ave', 'Orlando', 'FL', 'USA', '32801', '+1 (407) 999-7788', None, 'hleacock@gmail.com', 4),
 (23, 'John', 'Gordon', None, '69 Salem Street', 'Boston', 'MA', 'USA', '2113', '+1 (617) 522-1333', None, 'johngordon22@yahoo.com', 4),
 (24, 'Frank', 'Ralston', None, '162 E Superior Street', 'Chicago', 'IL', 'USA', '60611', '+1 (312) 332-3232', None, 'fralston@gmail.com', 3),
 (25, 'Victor', 'Stevens', None, '319 N. Frances Street', 'Madison', 'WI', 'USA', '53703', '+1 (608) 257-0597', None, 'vstevens@yahoo.com', 5)]

Now that we've got a way of pulling just US customers, we'll use a chain of joins to make sure we count only US customers.

In [10]:
%%sql
WITH
us_customers AS (
                SELECT *
                FROM customer
                WHERE country = "USA"
                ),
us_invoices AS (
                SELECT *
                FROM invoice i
                INNER JOIN us_customers us ON i.customer_id = us.customer_id
                ),
us_invoice_line AS 
                (
                SELECT *
                FROM invoice_line il
                INNER JOIN us_invoices ui ON il.invoice_id = ui.invoice_id
                ),
us_track AS
                (
                SELECT *
                FROM track t
                INNER JOIN us_invoice_line uil ON t.track_id = uil.track_id
                )
SELECT COUNT(track_id) FROM us_track;

 * sqlite:///chinook.db
Done.


[(1051,)]

Importantly, we first needed to check how many tracks were purchased by US customers, which we will be able to use going forward.

In [11]:
%%sql
WITH
us_customers AS (
                SELECT *
                FROM customer
                WHERE country = "USA"
                ),
us_invoices AS (
                SELECT *
                FROM invoice i
                INNER JOIN us_customers us ON i.customer_id = us.customer_id
                ),
us_invoice_line AS 
                (
                SELECT *
                FROM invoice_line il
                INNER JOIN us_invoices ui ON il.invoice_id = ui.invoice_id
                ),
us_track AS
                (
                SELECT *
                FROM track t
                INNER JOIN us_invoice_line uil ON t.track_id = uil.track_id
                )
SELECT
    g.name genre,
    COUNT(ut.track_id) track_count,
    COUNT(ut.track_id)/1051. track_pct
FROM genre g
INNER JOIN us_track ut ON ut.genre_id = g.genre_id
GROUP BY 1
ORDER BY 3 DESC;

 * sqlite:///chinook.db
Done.


[('Rock', 561, 0.5337773549000951),
 ('Alternative & Punk', 130, 0.12369172216936251),
 ('Metal', 124, 0.11798287345385347),
 ('R&B/Soul', 53, 0.05042816365366318),
 ('Blues', 36, 0.03425309229305423),
 ('Alternative', 35, 0.03330161750713606),
 ('Pop', 22, 0.02093244529019981),
 ('Latin', 22, 0.02093244529019981),
 ('Hip Hop/Rap', 20, 0.019029495718363463),
 ('Jazz', 14, 0.013320647002854425),
 ('Easy Listening', 13, 0.012369172216936251),
 ('Reggae', 6, 0.005708848715509039),
 ('Electronica/Dance', 5, 0.004757373929590866),
 ('Classical', 4, 0.003805899143672693),
 ('Heavy Metal', 3, 0.0028544243577545195),
 ('Soundtrack', 2, 0.0019029495718363464),
 ('TV Shows', 1, 0.0009514747859181732)]

So, if our new label has hip-hop, punk, pop, and blues, but can only choose 3 genres to get behind, we recommend punk, blues, and pop from this data. It's also worth mentioning that from this data, not having a rock artist appears to be a missed opportunity.

In [12]:
%%sql
SELECT e.*, SUM(i.total)
FROM employee e
INNER JOIN customer c on e.employee_id = c.support_rep_id
INNER JOIN invoice i on c.customer_id = i.customer_id
GROUP BY e.employee_id;

 * sqlite:///chinook.db
Done.


[(3, 'Peacock', 'Jane', 'Sales Support Agent', 2, '1973-08-29 00:00:00', '2017-04-01 00:00:00', '1111 6 Ave SW', 'Calgary', 'AB', 'Canada', 'T2P 5M5', '+1 (403) 262-3443', '+1 (403) 262-6712', 'jane@chinookcorp.com', 1731.5100000000039),
 (4, 'Park', 'Margaret', 'Sales Support Agent', 2, '1947-09-19 00:00:00', '2017-05-03 00:00:00', '683 10 Street SW', 'Calgary', 'AB', 'Canada', 'T2P 5G3', '+1 (403) 263-4423', '+1 (403) 263-4289', 'margaret@chinookcorp.com', 1584.0000000000034),
 (5, 'Johnson', 'Steve', 'Sales Support Agent', 2, '1965-03-03 00:00:00', '2017-10-17 00:00:00', '7727B 41 Ave', 'Calgary', 'AB', 'Canada', 'T3B 1Y7', '1 (780) 836-9987', '1 (780) 836-9543', 'steve@chinookcorp.com', 1393.920000000002)]

We can see that Jane Peacock has the most in sales, followed by Margaret Park, followed by Steve Johnson. However, it is worth noting, that is also the order of their hire dates, from oldest to newest. So it is possible this is not indicative of the quality of the employee, but how long they have been working there.

In [13]:
%%sql
    SELECT country, COUNT(customer_id) customer_count
    FROM customer
    GROUP BY country

 * sqlite:///chinook.db
Done.


[('Argentina', 1),
 ('Australia', 1),
 ('Austria', 1),
 ('Belgium', 1),
 ('Brazil', 5),
 ('Canada', 8),
 ('Chile', 1),
 ('Czech Republic', 2),
 ('Denmark', 1),
 ('Finland', 1),
 ('France', 5),
 ('Germany', 4),
 ('Hungary', 1),
 ('India', 2),
 ('Ireland', 1),
 ('Italy', 1),
 ('Netherlands', 1),
 ('Norway', 1),
 ('Poland', 1),
 ('Portugal', 2),
 ('Spain', 1),
 ('Sweden', 1),
 ('USA', 13),
 ('United Kingdom', 3)]

First thing I wanted to do was just to get customer counts for each country so I could get a sense of the data.

In [14]:
%%sql
WITH
customer_counts as (
    SELECT country, COUNT(customer_id) customer_count
    FROM customer
    GROUP BY country
    )
SELECT CASE
    WHEN customer_count > 1 THEN country
    ELSE "Other"
    END AS country_other,
    SUM(customer_count) customer_count_other
FROM customer_counts
GROUP BY country_other;

 * sqlite:///chinook.db
Done.


[('Brazil', 5),
 ('Canada', 8),
 ('Czech Republic', 2),
 ('France', 5),
 ('Germany', 4),
 ('India', 2),
 ('Other', 15),
 ('Portugal', 2),
 ('USA', 13),
 ('United Kingdom', 3)]

From there, I was able to make a new column that would either be the country's name or other, depending on whether they had more than one customer. I used this column to group by, so that all the others would count together.

In [15]:
%%sql
WITH
customer_counts as (
    SELECT country, COUNT(customer_id) customer_count
    FROM customer
    GROUP BY country
    ),
other_grouped as (
    SELECT 
        CASE
            WHEN customer_count > 1 THEN country
            ELSE "Other"
            END AS country_other,
        SUM(customer_count) customer_count
    FROM customer_counts
    GROUP BY country_other
    )
SELECT
    country_other,
    customer_count
FROM
    (
    SELECT *,
        CASE
            WHEN country_other = "Other" THEN 1
            ELSE 0
        END AS sort
    FROM other_grouped
    )
ORDER BY sort ASC;

 * sqlite:///chinook.db
Done.


[('Brazil', 5),
 ('Canada', 8),
 ('Czech Republic', 2),
 ('France', 5),
 ('Germany', 4),
 ('India', 2),
 ('Portugal', 2),
 ('USA', 13),
 ('United Kingdom', 3),
 ('Other', 15)]

In [16]:
%%sql
SELECT
        c.country,
        i.customer_id,
        COUNT(i.customer_id) order_count,
        SUM(i.total) total
    FROM customer c
    INNER JOIN invoice i ON c.customer_id = i.customer_id
    GROUP BY 2

 * sqlite:///chinook.db
Done.


[('Brazil', 1, 13, 108.89999999999998),
 ('Germany', 2, 11, 82.17),
 ('Canada', 3, 9, 99.99),
 ('Norway', 4, 9, 72.27000000000001),
 ('Czech Republic', 5, 18, 144.54000000000002),
 ('Czech Republic', 6, 12, 128.7),
 ('Austria', 7, 9, 69.3),
 ('Belgium', 8, 7, 60.38999999999999),
 ('Denmark', 9, 10, 37.61999999999999),
 ('Brazil', 10, 12, 60.39),
 ('Brazil', 11, 10, 69.3),
 ('Brazil', 12, 11, 82.17),
 ('Brazil', 13, 15, 106.91999999999999),
 ('Canada', 14, 10, 29.699999999999996),
 ('Canada', 15, 9, 66.33),
 ('USA', 16, 8, 74.25),
 ('USA', 17, 12, 98.01),
 ('USA', 18, 8, 79.2),
 ('USA', 19, 9, 54.449999999999996),
 ('USA', 20, 12, 95.03999999999999),
 ('USA', 21, 11, 91.08000000000001),
 ('USA', 22, 12, 92.07000000000001),
 ('USA', 23, 10, 66.33),
 ('USA', 24, 8, 71.28),
 ('USA', 25, 10, 76.23),
 ('USA', 26, 12, 86.13000000000002),
 ('USA', 27, 9, 84.14999999999999),
 ('USA', 28, 10, 72.27),
 ('Canada', 29, 4, 40.59),
 ('Canada', 30, 13, 91.08),
 ('Canada', 31, 11, 62.370000000000005),
 ('Canada', 32, 8, 70.28999999999999),
 ('Canada', 33, 12, 75.24000000000001),
 ('Portugal', 34, 13, 102.96000000000001),
 ('Portugal', 35, 16, 82.17),
 ('Germany', 36, 11, 85.14),
 ('Germany', 37, 10, 94.05000000000001),
 ('Germany', 38, 9, 73.25999999999999),
 ('France', 39, 9, 79.2),
 ('France', 40, 9, 72.27),
 ('France', 41, 9, 64.35),
 ('France', 42, 11, 99.99),
 ('France', 43, 12, 73.25999999999999),
 ('Finland', 44, 11, 79.2),
 ('Hungary', 45, 10, 78.21),
 ('Ireland', 46, 13, 114.83999999999997),
 ('Italy', 47, 9, 50.49),
 ('Netherlands', 48, 10, 65.34),
 ('Poland', 49, 10, 76.22999999999999),
 ('Spain', 50, 11, 98.01),
 ('Sweden', 51, 10, 75.24),
 ('United Kingdom', 52, 8, 68.31),
 ('United Kingdom', 53, 11, 98.01),
 ('United Kingdom', 54, 9, 79.2),
 ('Australia', 55, 10, 81.18),
 ('Argentina', 56, 5, 39.6),
 ('Chile', 57, 13, 97.02000000000001),
 ('India', 58, 13, 111.86999999999999),
 ('India', 59, 8, 71.28)]

In [17]:
%%sql
WITH
customer_counts as (
    SELECT country, COUNT(customer_id) customer_count
    FROM customer
    GROUP BY country
    ),
country_list AS (
    SELECT 
        country
    FROM customer_counts
    WHERE customer_count>1
),
customer_invoice AS (
    SELECT
        CASE
            WHEN c.country IN country_list THEN c.country
            ELSE "Other"
            END AS country,
        i.customer_id,
        COUNT(i.customer_id) order_count,
        SUM(i.total) total
    FROM customer c
    INNER JOIN invoice i ON c.customer_id = i.customer_id
    GROUP BY 2
    ),
main as (
    SELECT 
        country, 
        COUNT(customer_id) customer_count,
        SUM(total) total_sales,
        SUM(total)/COUNT(customer_id) avg_customer_value,
        SUM(total)/SUM(order_count) avg_order_value
    FROM customer_invoice ci
    GROUP BY country
    )
SELECT
    country, 
    customer_count, 
    ROUND(total_sales, 2) total_sales, 
    ROUND(avg_customer_value, 2) avg_customer_value,
    ROUND(avg_order_value, 2) avg_order_value
FROM
    (
    SELECT *,
        CASE
            WHEN country = "Other" THEN 1
            ELSE 0
        END AS sort
    FROM main
    )
ORDER BY sort ASC;

 * sqlite:///chinook.db
Done.


[('Brazil', 5, 427.68, 85.54, 7.01),
 ('Canada', 8, 535.59, 66.95, 7.05),
 ('Czech Republic', 2, 273.24, 136.62, 9.11),
 ('France', 5, 389.07, 77.81, 7.78),
 ('Germany', 4, 334.62, 83.66, 8.16),
 ('India', 2, 183.15, 91.58, 8.72),
 ('Portugal', 2, 185.13, 92.57, 6.38),
 ('USA', 13, 1040.49, 80.04, 7.94),
 ('United Kingdom', 3, 245.52, 81.84, 8.77),
 ('Other', 15, 1094.94, 73.0, 7.45)]

We can glean all sorts of useful information from this. We can see that the US and "Other", ie smaller countries, are the countries with the most customers. But we can also see that the Czech Republic spends a lot per customer and per order. The UK and India also spend a lot per order.

# Album vs Individual Song Purchases

In [18]:
%%sql
SELECT * FROM invoice_line
LIMIT 50;

 * sqlite:///chinook.db
Done.


[(1, 1, 1158, 0.99, 1),
 (2, 1, 1159, 0.99, 1),
 (3, 1, 1160, 0.99, 1),
 (4, 1, 1161, 0.99, 1),
 (5, 1, 1162, 0.99, 1),
 (6, 1, 1163, 0.99, 1),
 (7, 1, 1164, 0.99, 1),
 (8, 1, 1165, 0.99, 1),
 (9, 1, 1166, 0.99, 1),
 (10, 1, 1167, 0.99, 1),
 (11, 1, 1168, 0.99, 1),
 (12, 1, 1169, 0.99, 1),
 (13, 1, 1170, 0.99, 1),
 (14, 1, 1171, 0.99, 1),
 (15, 1, 1172, 0.99, 1),
 (16, 1, 1173, 0.99, 1),
 (17, 2, 3476, 0.99, 1),
 (18, 2, 482, 0.99, 1),
 (19, 2, 2701, 0.99, 1),
 (20, 2, 1641, 0.99, 1),
 (21, 2, 1119, 0.99, 1),
 (22, 2, 2324, 0.99, 1),
 (23, 2, 201, 0.99, 1),
 (24, 2, 819, 0.99, 1),
 (25, 2, 392, 0.99, 1),
 (26, 2, 2651, 0.99, 1),
 (27, 3, 2516, 0.99, 1),
 (28, 3, 2646, 0.99, 1),
 (29, 4, 3448, 0.99, 1),
 (30, 4, 2560, 0.99, 1),
 (31, 4, 3336, 0.99, 1),
 (32, 4, 829, 0.99, 1),
 (33, 4, 1872, 0.99, 1),
 (34, 4, 748, 0.99, 1),
 (35, 4, 1778, 0.99, 1),
 (36, 4, 2514, 0.99, 1),
 (37, 5, 1986, 0.99, 1),
 (38, 5, 1987, 0.99, 1),
 (39, 5, 1988, 0.99, 1),
 (40, 5, 1989, 0.99, 1),
 (41, 5, 1990, 0.99, 1),
 (42, 5, 1991, 0.99, 1),
 (43, 5, 1992, 0.99, 1),
 (44, 5, 1993, 0.99, 1),
 (45, 5, 1994, 0.99, 1),
 (46, 5, 1995, 0.99, 1),
 (47, 5, 1996, 0.99, 1),
 (48, 5, 1997, 0.99, 1),
 (49, 5, 1998, 0.99, 1),
 (50, 5, 1999, 0.99, 1)]

I wanted to get a sense of what the invoice line data looks like.

I am going to use the Group_Concat command to basically concatinate all track_ids on a single invoice into one list. From there, I will check the list to see if it matches any albums.

In [23]:
%%sql
CREATE VIEW invoice_tracks AS
SELECT invoice_id, GROUP_CONCAT(track_id) tracks_purchased, COUNT(track_id) track_count
FROM invoice_line
GROUP BY 1

 * sqlite:///chinook.db
(sqlite3.OperationalError) table invoice_tracks already exists
[SQL: CREATE VIEW invoice_tracks AS
SELECT invoice_id, GROUP_CONCAT(track_id) tracks_purchased, COUNT(track_id) track_count
FROM invoice_line
GROUP BY 1]
(Background on this error at: http://sqlalche.me/e/14/e3q8)


Since I anticipate using these tables a lot, I am going to create a view so that I can reference them easily, as opposed to using with statements over and over.

In [24]:
%%sql
SELECT * FROM invoice_tracks;

 * sqlite:///chinook.db
Done.


[(1, '1158,1159,1160,1161,1162,1163,1164,1165,1166,1167,1168,1169,1170,1171,1172,1173', 16),
 (2, '3476,482,2701,1641,1119,2324,201,819,392,2651', 10),
 (3, '2516,2646', 2),
 (4, '3448,2560,3336,829,1872,748,1778,2514', 8),
 (5, '1986,1987,1988,1989,1990,1991,1992,1993,1994,1995,1996,1997,1998,1999,2000,2001,2002', 17),
 (6, '1045,30', 2),
 (7, '3477,3439,1102,711,1988,3024,834,2596,1343,2644,42', 11),
 (8, '2231,2509,81,1758,1455,2946,574,1752,2263,1560', 10),
 (9, '206,3456,1041,1279,2098,3399,1983,993,196', 9),
 (10, '2663,3290', 2),
 (11, '2597,2203,2420,2643,3072,1138,2391,2257,610,1419,3452', 11),
 (12, '1463,690,2663,92', 4),
 (13, '2553', 1),
 (14, '541', 1),
 (15, '3050,1489,807,3336', 4),
 (16, '2160,875,83,16,1756,3460,93,24,21', 9),
 (17, '2400,1610,1973,2553,59,2270,55,2564,2171,2522,1419', 11),
 (18, '1027,3142,1583,1752,2736', 5),
 (19, '105,2669,1784,3460,2640,1310,2510', 7),
 (20, '905,102,3336,1132,337,60,3357,3225,3297,3297', 10),
 (21, '1461,2556,1434,3135,1473,1502,13,2971,2561', 9),
 (22, '1000,383', 2),
 (23, '1,6,7,8,9,10,11,12,13,14', 10),
 (24, '1146,1147,1148,1149,1150,1151,1152,1153,1154,1155,1156,1157', 12),
 (25, '46,948,2185,3049,1167,32,1800,1477', 8),
 (26, '2670,85', 2),
 (27, '3439,3014,1280,1496,2229,1445,2230,3105', 8),
 (28, '36,3456,932', 3),
 (29, '1025,2462,3341,1626,2395,479', 6),
 (30, '919,3461,33,2341,1139,3109,2013,990,2413,2396,3464', 11),
 (31, '2731,2732,2733,2734,2735,2736,2737,2738,2739,2740,2741,2742,2743,2744,2745,2746,2747,2748,2749,2750', 20),
 (32, '38,39,40,41,42,43,44,45,46,47,48,49,50', 13),
 (33, '3307,2736,2542,2697', 4),
 (34, '610', 1),
 (35, '2377,80,836,6,3475', 5),
 (36, '1582,968,2233,153,1493,2656,1168', 7),
 (37, '3458,1428,3329,698,1014,3336,3297,1166,3392,1763', 10),
 (38, '2960,154,3272,2557,1566,1472,3409', 7),
 (39, '2308,1669,2415,3142,3366,184,2519,3424', 8),
 (40, '2382,2654,2556,108,553,2737,686,1626,3460,2237', 10),
 (41, '153,3457,2511,2698,2945,1947,3336,773', 8),
 (42, '2646,3074,711,13,1047,699,580,897,153', 9),
 (43, '918,2342,1162,1654,1768,2641,1997', 7),
 (44, '3052,3053,3054,3055,3056,3057,3058,3059,3060,3061,3062,3063', 12),
 (45, '1890,1787,2731,2649,102,158,1976,1837', 8),
 (46, '2639,2640,2641,2642,2643,2644,2645,2646,2647,2648,2649', 11),
 (47, '3396,2518,2095,449,478,8', 6),
 (48, '2210,427,1167,58', 4),
 (49, '2561,1877,146,1162,146', 5),
 (50, '1905,2155,3388,2455,1794,1511,3455', 7),
 (51, '2265,2171,692,3046,1184,126,475,34', 8),
 (52, '2664,2665,2666,2667,2668,2669,2670,2671,2672,2673,2674,2675', 12),
 (53, '1551,207,1846', 3),
 (54, '1773,1774,1775,1776,1777,1778,1779,1780,1781,1782,1783,1784,1785,1786,1787,1788,1789,1790', 18),
 (55, '1552,2672,2006,2453,548,349,2552,2675,155,481', 10),
 (56, '2254,2255,2256,2257,2258,2259,2260,2261,2262,2263,2264,2265,2266,2267,2268,2269,2270', 17),
 (57, '2948', 1),
 (58, '1029,2644,2395,2669,2630,1745', 6),
 (59, '449,3491,1053,2640,2563,2731,1123,198,3336,1463,3390', 11),
 (60, '479,1875,544,1030,2416,1167,1826,46', 8),
 (61, '2609,2610,2611,2612,2613,2614,2615,2616,2617,2618,2619,2620', 12),
 (62, '2989,97,2680', 3),
 (63, '444,3470,3376,1994,6,8', 6),
 (64, '38,39,40,41,42,43,44,45,46,47,48,49,50', 13),
 (65, '1139,539', 2),
 (66, '1435,1761,3377,1785,442,1951,2334', 7),
 (67, '2401,2287,84,3099,3468,2413,2164,3478,787,9,861', 11),
 (68, '2575,1669,2265,61,751,3353,2008,1903,910,2236,1990', 11),
 (69, '85,86,87,88,89,90,91,92,93,94,95,96,97,98', 14),
 (70, '2307', 1),
 (71, '2254,2255,2256,2257,2258,2259,2260,2261,2262,2263,2264,2265,2266,2267,2268,2269,2270', 17),
 (72, '30,2564,1658', 3),
 (73, '2391,2392,2393,2394,2395,2396,2397,2398,2399,2400,2401,2402,2403,2404,2405', 15),
 (74, '1743,3081', 2),
 (75, '22,1488,1977,1233', 4),
 (76, '1020,1021,1022,1023,1024,1025,1026,1027,1028,1029,1030,1031,1032', 13),
 (77, '15,17,18,19,20,22', 6),
 (78, '2592,1052,33,3336,1215,3370,2553,1126,3006', 9),
 (79, '1902,1903,1904,1905,1906,1907,1908,1909,1910,1911,1912,1913,1914,1915', 14),
 (80, '3400'

By using Group_Concat, we can see all the songs purchased in each invoice.

In [25]:
%%sql
CREATE VIEW album_tracks AS
SELECT album_id, GROUP_CONCAT(track_id) tracks_featured, COUNT(track_id) track_count
FROM track
GROUP BY 1;

 * sqlite:///chinook.db
(sqlite3.OperationalError) table album_tracks already exists
[SQL: CREATE VIEW album_tracks AS
SELECT album_id, GROUP_CONCAT(track_id) tracks_featured, COUNT(track_id) track_count
FROM track
GROUP BY 1;]
(Background on this error at: http://sqlalche.me/e/14/e3q8)


In [26]:
%%sql
SELECT * FROM album_tracks;

 * sqlite:///chinook.db
Done.


[(1, '1,6,7,8,9,10,11,12,13,14', 10),
 (2, '2', 1),
 (3, '3,4,5', 3),
 (4, '15,16,17,18,19,20,21,22', 8),
 (5, '23,24,25,26,27,28,29,30,31,32,33,34,35,36,37', 15),
 (6, '38,39,40,41,42,43,44,45,46,47,48,49,50', 13),
 (7, '51,52,53,54,55,56,57,58,59,60,61,62', 12),
 (8, '63,64,65,66,67,68,69,70,71,72,73,74,75,76', 14),
 (9, '77,78,79,80,81,82,83,84', 8),
 (10, '85,86,87,88,89,90,91,92,93,94,95,96,97,98', 14),
 (11, '99,100,101,102,103,104,105,106,107,108,109,110', 12),
 (12, '111,112,113,114,115,116,117,118,119,120,121,122', 12),
 (13, '123,124,125,126,127,128,129,130', 8),
 (14, '131,132,133,134,135,136,137,138,139,140,141,142,143', 13),
 (15, '144,145,146,147,148', 5),
 (16, '149,150,151,152,153,154,155', 7),
 (17, '156,157,158,159,160,161,162,163,164,165', 10),
 (18, '166,167,168,169,170,171,172,173,174,175,176,177,178,179,180,181,182', 17),
 (19, '183,184,185,186,187,188,189,190,191,192,193', 11),
 (20, '194,195,196,197,198,199,200,201,202,203,204', 11),
 (21, '205,206,207,208,209,210,211,212,213,214,215,216,217,218,219,220,221,222', 18),
 (22, '223,224,225', 3),
 (23, '226,227,228,229,230,231,232,233,234,235,236,237,238,239,240,241,242,243,244,245,515,516,517,518,519,520,521,522,523,524,525,526,527,528', 34),
 (24, '246,247,248,249,250,251,252,253,254,255,256,257,258,259,260,261,262,263,264,265,266,267,268', 23),
 (25, '269,270,271,272,273,274,275,276,277,278,279,280,281', 13),
 (26, '282,283,284,285,286,287,288,289,290,291,292,293,294,295,296,297,298', 17),
 (27, '299,300,301,302,303,304,305,306,307,308,309,310,311,312', 14),
 (28, '313,314,315,316,317,318,319,320,321,322', 10),
 (29, '323,324,325,326,327,328,329,330,331,332,333,334,335,336', 14),
 (30, '337,338,339,340,341,342,343,344,345,346,347,348,349,350', 14),
 (31, '351,352,353,354,355,356,357,358,359', 9),
 (32, '360,361,362,363,364,365,366,367,368,369,370,371,372,373', 14),
 (33, '374,375,376,377,378,379,380,381,382,383,384,385,386,387,388,389,390', 17),
 (34, '391,392,393,394,395,396,397,398,399,400,401,402,403,404,405,406,407', 17),
 (35, '408,409,410,411,412,413,414,415,416,417,418', 11),
 (36, '419,420,421,422,423,424,425,426,427,428,429,430,431,432,433,434,435', 17),
 (37, '436,437,438,439,440,441,442,443,444,445,446,447,448,449,450,451,452,453,454,455', 20),
 (38, '456,457,458,459,460,461,462,463,464,465,466,467', 12),
 (39, '468,469,470,471,472,473,474,475,476,477,478,479,480,481,482,483,484,485,486,487,488', 21),
 (40, '489,490,491,492,493,494,495,496,497,498,499,500', 12),
 (41, '501,502,503,504,505,506,507,508,509,510,511,512,513,514', 14),
 (42, '529,530,531,532,533,534,535,536,537,538,539,540,541,542', 14),
 (43, '543,544,545,546,547,548,549', 7),
 (44, '550,551,552,553,554,555', 6),
 (45, '556,557,558,559,560,561,562,563,564,565,566,567,568,569', 14),
 (46, '570,571,572,573,574,575,576,577,578,579,580,581,582', 13),
 (47, '583,584,585,586,587,588,589,590,591,592,593,594,595,596', 14),
 (48, '597,598,599,600,601,602,603,604,605,606,607,608,609', 13),
 (49, '610,611,612,613,614,615,616,617,618,619', 10),
 (50, '620,621,622,623', 4),
 (51, '624,625,626,627,628,629,630,631,632,633,634,635,636,637,638,639,640,641,642,643,644,645', 22),
 (52, '646,647,648,649,650,651,652,653,654,655,656,657,658,659,660', 15),
 (53, '661,662,663,664,665,666,667,668,669,670,671,672,673,674', 14),
 (54, '675,676,677,678,679,680,681,682,683,684,685,686,687,688,689,690,691,692,693,694', 20),
 (55, '695,696,697,698,699,700,701,702,703,704,705,706,707,708,709,710,711,712,713,714', 20),
 (56, '715,716,717,718,719,720,721,722,723,724,725,726,727,728,729', 15),
 (57, '730,731,732,733,734,735,736,737,738,739,740,741,742,743,744', 15),
 (58, '745,746,747,748,749,750,751,752,753', 9),
 (59, '754,755,756,757,758,759,760', 7),
 (60, '761,762,763,764,765,766,767', 7),
 (61, '768,769,770,771,772,773,774,775,776,777,778', 11),
 (62, '779,780,781,782,783,784,785', 7),
 (63, '786,787,788,789,790,791,792,793,794,795,796,797', 12),
 (64, '798,799,800,801,802,803,804,805,806', 9),
 (65, '807,808

This shows all tracks featured on an album. Now I need to check each album to see if it shows up in an invoice in its entirety. I am also going to eliminate any albums that have 3 or less songs.

In [29]:
%%sql
WITH
large_album_tracks AS
    (
    SELECT *
    FROM album_tracks
    WHERE track_count > 3
    )
SELECT lat.album_id, it.*
FROM invoice_tracks it
INNER JOIN large_album_tracks lat ON lat.tracks_featured = it.tracks_purchased;

 * sqlite:///chinook.db
Done.


[(1, 23, '1,6,7,8,9,10,11,12,13,14', 10),
 (1, 414, '1,6,7,8,9,10,11,12,13,14', 10),
 (4, 200, '15,16,17,18,19,20,21,22', 8),
 (4, 230, '15,16,17,18,19,20,21,22', 8),
 (5, 142, '23,24,25,26,27,28,29,30,31,32,33,34,35,36,37', 15),
 (5, 427, '23,24,25,26,27,28,29,30,31,32,33,34,35,36,37', 15),
 (6, 32, '38,39,40,41,42,43,44,45,46,47,48,49,50', 13),
 (6, 64, '38,39,40,41,42,43,44,45,46,47,48,49,50', 13),
 (7, 300, '51,52,53,54,55,56,57,58,59,60,61,62', 12),
 (8, 532, '63,64,65,66,67,68,69,70,71,72,73,74,75,76', 14),
 (9, 262, '77,78,79,80,81,82,83,84', 8),
 (10, 69, '85,86,87,88,89,90,91,92,93,94,95,96,97,98', 14),
 (16, 465, '149,150,151,152,153,154,155', 7),
 (16, 511, '149,150,151,152,153,154,155', 7),
 (20, 284, '194,195,196,197,198,199,200,201,202,203,204', 11),
 (36, 479, '419,420,421,422,423,424,425,426,427,428,429,430,431,432,433,434,435', 17),
 (37, 526, '436,437,438,439,440,441,442,443,444,445,446,447,448,449,450,451,452,453,454,455', 20),
 (42, 372, '529,530,531,532,533,534,535,536,537,538,539,540,541,542', 14),
 (55, 362, '695,696,697,698,699,700,701,702,703,704,705,706,707,708,709,710,711,712,713,714', 20),
 (65, 512, '807,808,809,810,811,812,813,814,815', 9),
 (72, 582, '891,892,893,894,895,896,897,898,899,900,901,902,903,904,905,906,907,908', 18),
 (80, 388, '999,1000,1001,1002,1003,1004,1005,1006,1007,1008', 10),
 (80, 420, '999,1000,1001,1002,1003,1004,1005,1006,1007,1008', 10),
 (82, 76, '1020,1021,1022,1023,1024,1025,1026,1027,1028,1029,1030,1031,1032', 13),
 (83, 183, '1033,1034,1035,1036,1037,1038,1039,1040,1041,1042,1043,1044,1045,1046,1047,1048,1049,1050,1051,1052,1053,1054,1055,1056', 24),
 (88, 212, '1121,1122,1123,1124,1125,1126,1127,1128,1129,1130,1131,1132', 12),
 (88, 253, '1121,1122,1123,1124,1125,1126,1127,1128,1129,1130,1131,1132', 12),
 (89, 330, '1133,1134,1135,1136,1137,1138,1139,1140,1141,1142,1143,1144,1145', 13),
 (89, 381, '1133,1134,1135,1136,1137,1138,1139,1140,1141,1142,1143,1144,1145', 13),
 (90, 24, '1146,1147,1148,1149,1150,1151,1152,1153,1154,1155,1156,1157', 12),
 (90, 399, '1146,1147,1148,1149,1150,1151,1152,1153,1154,1155,1156,1157', 12),
 (91, 1, '1158,1159,1160,1161,1162,1163,1164,1165,1166,1167,1168,1169,1170,1171,1172,1173', 16),
 (92, 404, '1174,1175,1176,1177,1178,1179,1180,1181,1182,1183,1184,1185,1186,1187', 14),
 (100, 375, '1268,1269,1270,1271,1272,1273,1274,1275,1276', 9),
 (105, 196, '1325,1326,1327,1328,1329,1330,1331,1332,1333,1334', 10),
 (109, 590, '1362,1363,1364,1365,1366,1367,1368,1369,1370', 9),
 (115, 92, '1414,1415,1416,1417,1418,1419,1420,1421,1422,1423,1424,1425,1426,1427,1428,1429,1430,1431,1432,1433', 20),
 (116, 485, '1434,1435,1436,1437,1438,1439,1440,1441,1442,1443', 10),
 (118, 188, '1455,1456,1457,1458,1459,1460,1461,1462,1463,1464,1465', 11),
 (119, 146, '1466,1467,1468,1469,1470,1471,1472,1473,1474,1475,1476,1477,1478', 13),
 (119, 171, '1466,1467,1468,1469,1470,1471,1472,1473,1474,1475,1476,1477,1478', 13),
 (119, 296, '1466,1467,1468,1469,1470,1471,1472,1473,1474,1475,1476,1477,1478', 13),
 (119, 501, '1466,1467,1468,1469,1470,1471,1472,1473,1474,1475,1476,1477,1478', 13),
 (120, 112, '1479,1480,1481,1482,1483,1484,1485,1486,1487,1488,1489,1490,1491,1492,1493,1494,1495', 17),
 (120, 114, '1479,1480,1481,1482,1483,1484,1485,1486,1487,1488,1489,1490,1491,1492,1493,1494,1495', 17),
 (120, 273, '1479,1480,1481,1482,1483,1484,1485,1486,1487,1488,1489,1490,1491,1492,1493,1494,1495', 17),
 (120, 344, '1479,1480,1481,1482,1483,1484,1485,1486,1487,1488,1489,1490,1491,1492,1493,1494,1495', 17),
 (120, 449, '1479,1480,1481,1482,1483,1484,1485,1486,1487,1488,1489,1490,1491,1492,1493,1494,1495', 17),
 (120, 478, '1479,1480,1481,1482,1483,1484,1485,1486,1487,1488,1489,1490,1491,1492,1493,1494,1495', 17),
 (123, 341, '1520,1521,1522,1523,1524,1525,1526,1527,1528,1529,1530,1531', 12),
 (126, 457, '1562,1563,1564,1565,1566,1567,1568,1569,1570,1571,1572,1573,1574,1575,1576', 15),
 (145, 175, '1755,1756,1757,1758,1759,1760,1761,1762,1763,1764,1765,1766,1767,1768,176

Success! We have all invoices that feature solely the album at hand! Now we just need to count them, and them compare that to the total number of songs purchased.

In [33]:
%%sql
WITH
large_album_tracks AS
    (
    SELECT *
    FROM album_tracks
    WHERE track_count > 3
    ),
album_invoice AS
    (
    SELECT lat.album_id, it.*
    FROM invoice_tracks it
    INNER JOIN large_album_tracks lat ON lat.tracks_featured = it.tracks_purchased
    )
SELECT COUNT(album_id)
FROM album_invoice;

 * sqlite:///chinook.db
Done.


[(110,)]

Great! We know there are 110 invoices that are an album. Now we just need to know the total amount of invoices to find the percentage.

In [36]:
%%sql
SELECT COUNT(DISTINCT invoice_id)
FROM invoice_line;

 * sqlite:///chinook.db
Done.


[(614,)]

There are 614 total invoices. That means that if 110 were album purchases, 514 were not. Seems like albums make up a small percentage of the total purchases.